### PR TITLE
G-API: Implement text mask blending for render

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@smirnov-alexey @dmatveev

**Comment by TolyaTalamanov** [comment]
@smirnov-alexey @dmatveev

File: modules/gapi/src/api/render_ocv.cpp [hunk 1]
```diff
@@ -222,5 +225,73 @@ void drawPrimitivesOCVYUV(cv::Mat &in, const Prims &prims)
...
+                  (mask.rows + mask_size.height - *baseline) / 2);
+
+    cv::putText(mask, text_prim.text, org, text_prim.ff,
+                text_prim.fs, 255, text_prim.thick);
+}
+
+void cv::gapi::ocv::blendImage(const cv::Mat& img,
+                                     const cv::Mat& alpha,
+                                     const cv::Point& org,
+                                     cv::Mat background)
```
Comment (smirnov-alexey) [inline modules/gapi/src/api/render_ocv.cpp]: Indentation is broken a bit

File: modules/gapi/src/api/render_ocv.cpp [hunk 2]
```diff
@@ -222,5 +225,73 @@ void drawPrimitivesOCVYUV(cv::Mat &in, const Prims &prims)
...
+{
+    GAPI_Assert(img.type()   == CV_32FC3);
```
Comment (smirnov-alexey) [inline modules/gapi/src/api/render_ocv.cpp]: Are you sure it must be 32FC3? What's the problem with 8U?

File: modules/gapi/src/api/render_ocv.cpp [hunk 3]
```diff
@@ -222,5 +225,73 @@ void drawPrimitivesOCVYUV(cv::Mat &in, const Prims &prims)
 
 } // namespace draw
 } // namespace wip
+
+void cv::gapi::ocv::createBitMask(const cv::gapi::wip::draw::Text& text_prim,
+                                  cv::Mat& mask,
+                                  int* baseline) {
```
Comment (smirnov-alexey) [inline modules/gapi/src/api/render_ocv.cpp]: Why do you need a pointer to this?

File: modules/gapi/src/api/render_priv.hpp [hunk 4]
```diff
@@ -24,6 +24,24 @@ GAPI_EXPORTS void BGR2NV12(const cv::Mat& bgr, cv::Mat& y_plane, cv::Mat& uv_pla
...
+
+namespace ocv
+{
+    void createBitMask(const wip::draw::Text& text_prim,
+                       cv::Mat& mask,
+                       int* baseline);
+
+    // FIXME only for tests
+    GAPI_EXPORTS void blendImage(const cv::Mat& img,
+            const cv::Mat& alpha,
```
Comment (smirnov-alexey) [inline modules/gapi/src/api/render_priv.hpp]: Indentation

Comment (smirnov-alexey) [reply modules/gapi/src/api/render_priv.hpp]: I don't see any tests on that. Please, remove from the header or test properly (I would recommend removing).

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/src/api/render_ocv.cpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_ocv.cpp]: Fixed

File: modules/gapi/src/api/render_ocv.cpp [hunk 5]
```diff
@@ -222,5 +225,73 @@ void drawPrimitivesOCVYUV(cv::Mat &in, const Prims &prims)
...
+
+    cv::multiply(img32f, img32f_w, img32f);
+    cv::multiply(roi32f, roi32f_w, roi32f);
+    roi32f += img32f;
+
+    roi32f.convertTo(roi, CV_8U, 255.0);
+}
+
+void cv::gapi::ocv::putText(const cv::gapi::wip::draw::Text& text_p,
+                                          cv::Mat& background)
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_ocv.cpp]: Fixed

File: modules/gapi/src/api/render_priv.hpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_priv.hpp]: Removed export

File: modules/gapi/src/api/ocv_mask_creator.hpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+//
+// Copyright (C) 2018 Intel Corporation
+
+
+#ifndef OPENCV_OCV_MASK_CREATOR_HPP
+#define OPENCV_OCV_MASK_CREATOR_HPP
+
+#include "api/render_priv.hpp"
+
+namespace cv{
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/ocv_mask_creator.hpp]: Add space

File: modules/gapi/src/api/ocv_mask_creator.hpp [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+namespace gapi{
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/ocv_mask_creator.hpp]: Here too

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+#include <ft2build.h>
+#include FT_FREETYPE_H
+
+namespace cv {
+namespace gapi {
+namespace wip {
+namespace draw {
+
+class FreeTypeBitmaskCreator : public IBitmaskCreator {
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Move implementation to separate .cpp file

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+public:
+    FreeTypeBitmaskCreator()
+    {
+        FT_Init_FreeType(&m_library);
+
+        const char* font = std::getenv("FREETYPE_FONT");
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: What about CI ? Who will set this flag ?

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        return m_baseline;
+    }
+
+    void virtual setMaskParams(const cv::gapi::wip::draw::Text& text) override
+    {
+        m_text = text;
+
+        // Convert OpenCV scale to Freetype text height
+        auto sz = cv::getTextSize(text.text, text.ff, text.fs, text.thick, nullptr);
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: It is't the best approach to convert OpenCV scale to FreeType height font

File: modules/gapi/test/common/gapi_render_tests_inl.hpp [hunk 11]
```diff
@@ -39,6 +39,74 @@ TEST_P(RenderBGR, AccuracyTest)
     EXPECT_EQ(0, cv::norm(mat_gapi, mat_ocv));
 }
 
+TEST_P(RenderTextTestBGR, AccuracyTest)
```
Comment (TolyaTalamanov) [inline modules/gapi/test/common/gapi_render_tests_inl.hpp]: Maybe avoid duplicating code in the future

File: modules/gapi/src/api/render_priv.hpp [hunk 12]
```diff
@@ -22,6 +22,21 @@ namespace draw
...
+                const cv::Mat& alpha,
+                const cv::Point& org,
+                cv::Mat background);
+
+class IBitmaskCreator
+{
+public:
+    static std::unique_ptr<IBitmaskCreator> create(BackendT type);
+    virtual int createMask(cv::Mat&) = 0;
+    virtual const cv::Size& computeMaskSize() = 0;
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/render_priv.hpp]: Maybe just `cv::Size`

File: modules/gapi/src/api/render_ocv.cpp [hunk 13]
```diff
@@ -193,7 +210,8 @@ void drawPrimitivesOCV(cv::Mat &in, const Prims &prims)
                 cv::Mat img;
                 converter.cvtImg(i_p.img, img);
 
-                image(in, i_p.org, img, i_p.alpha);
```
Comment (TolyaTalamanov) [inline modules/gapi/src/api/render_ocv.cpp]: Remove this function

Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_ocv.cpp]: Fixed

File: modules/gapi/src/api/ocv_mask_creator.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/ocv_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/ocv_mask_creator.hpp (hunk 6 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/ocv_mask_creator.hpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev @smirnov-alexey @dbudniko

**Comment by dmatveev** [comment]
> This pullrequest changes

Changes what? Please write a full description

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 14]
```diff
@@ -43,7 +61,7 @@ struct Text
     cv::Scalar  color;              //!< The text color
     int         thick;              //!< The thickness of the lines used to draw a text
     int         lt;                 //!< The line type. See #LineTypes
-    bool        bottom_left_origin; //!< When true, the image data origin is at the bottom-left corner. Otherwise, it is at the top-left corner
+    BackendT    backend;            //!< Backend which create mask with text (OpenCV/FreeType)
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/render.hpp]: May be we could make it a CMake OPTION?

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/render.hpp]: For sure it needs to be discussed, maybe our tomorrow's forum is good for that

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 15]
```diff
@@ -0,0 +1,136 @@
+#ifdef HAVE_FREETYPE
+
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: In this case, 2019

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+        m_glyphs.resize(m_text.text.size());
+        m_pos.reserve(m_text.text.size());
+
+        for (size_t i = 0; i < m_text.text.size(); ++i)
+        {
+            FT_Load_Char(m_face, m_text.text[i], FT_LOAD_RENDER);
+            FT_Bitmap *bitmap = &(m_slot->bitmap);
+
+            cv::Mat(bitmap->rows, bitmap->width, CV_8UC1, bitmap->buffer, bitmap->pitch).copyTo(m_glyphs[i]);
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: so it renders the stuff while calculating the mask, right?

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            // FIXME why bitmap->width > shift m_slot->advance.x ?
+            if (shift < bitmap->width)
+            {
+                gl_x_pad = bmp_w;
+                bmp_w += bitmap->width;
+            }
+            else
+            {
+                bmp_w += shift;
+            }
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Please put a comment what the above code block does and why.

File: modules/gapi/src/api/render_ocv.cpp [hunk 18]
```diff
@@ -156,10 +133,29 @@ void drawPrimitivesOCV(cv::Mat &in, const Prims &prims)
...
+
+                mask.convertTo(mask, CV_32FC1, 1 / 255.0);
+                cv::Mat color_mask;
+
+                cv::merge(std::vector<cv::Mat>(3, mask), color_mask);
+                cv::Scalar color32f = t_p.color / 255.0;
+                cv::multiply(color_mask, color32f, color_mask);
+
+                // Org is bottom left point, trasform it to top left point for blendImage
+                cv::Point tl(t_p.org.x, t_p.org.y - color_mask.size().height + baseline);
```
Comment (dmatveev) [inline modules/gapi/src/api/render_ocv.cpp]: why 32f?

File: modules/gapi/src/api/render_ocv.cpp (hunk 2 above)
Comment (dmatveev) [reply modules/gapi/src/api/render_ocv.cpp]: +1

File: modules/gapi/src/api/render_priv.hpp (hunk 12 above)
Comment (dmatveev) [reply modules/gapi/src/api/render_priv.hpp]: it is safer to return cv::Size

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 14 above)
Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Agree with CMake option

Comment (alalek) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Conditional compilation should be avoided in public headers.

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+        return m_baseline;
+    }
+
+    void virtual setMaskParams(const cv::gapi::wip::draw::Text& text) override
+    {
+        m_text = text;
+
+        // Convert OpenCV scale to Freetype text height
+        auto sz = cv::getTextSize(text.text, text.ff, text.fs, text.thick, nullptr);
+        int font_height = static_cast<int>(sz.height * 3.0 / 2); // 3.0 / 2 for better aligning with OpenCV
```
Comment (smirnov-alexey) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Can't we do something with 3/2 considering the current approach?

File: modules/gapi/src/api/ocv_mask_creator.hpp [hunk 20]
```diff
@@ -0,0 +1,67 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
```
Comment (smirnov-alexey) [inline modules/gapi/src/api/ocv_mask_creator.hpp]: 2019

**Comment by dmatveev** [comment]
Is it worth reviewing or we need to wait while #15699 merges and this one gets updated?

File: modules/gapi/src/api/render_priv.hpp (hunk 12 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_priv.hpp]: Fixed

File: modules/gapi/src/api/ocv_mask_creator.hpp (hunk 20 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/ocv_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 15 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/include/opencv2/gapi/render.hpp (hunk 14 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Solved

File: modules/gapi/include/opencv2/gapi/render.hpp [hunk 21]
```diff
@@ -30,11 +30,29 @@ namespace wip
...
+    Text() = default;
+
+    Text(const std::string& text_,
+         const cv::Point& org_,
+         int ff_,
+         double fs_,
+         const cv::Scalar& color_,
+         int thick_ = 1,
+         int lt_ = LINE_8,
+         BackendT backend_ = BackendT::OpenCV)
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/render.hpp]: Solved

File: modules/gapi/src/api/render_ocv.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_ocv.cpp]: Removed this function

File: modules/gapi/src/api/render_ocv.cpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/render_ocv.cpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 16 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: No, it copy freetype glyph's to cv::Mat's

**Comment by dmatveev** [comment]
Where's the mandatory G-API standalone build mantra?

**Comment by TolyaTalamanov** [comment]
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    FT_Face       m_face;
+    FT_GlyphSlot  m_slot;
+
+    std::vector<cv::Mat> m_glyphs;
+    std::vector<cv::Point> m_pos;
+
+    cv::gapi::wip::draw::Text m_text;
+    int m_baseline;
+    int m_max_glyph_top;
+    cv::Size m_mask_size;
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/ocv_mask_creator.hpp [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        m_text = text;
+    }
+
+private:
+    cv::gapi::wip::draw::Text m_text;
+    int m_baseline;
+    int m_thick;
+    int m_ff;
+    cv::Size m_mask_size;
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/api/ocv_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+namespace draw
+{
+
+class FreeTypeBitmaskCreator : public IBitmaskCreator
+{
+public:
+    FreeTypeBitmaskCreator(const std::string& path)
+    {
+        FT_Init_FreeType(&m_library);
+        FT_New_Face(m_library, path.c_str(), 0, &m_face);
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 19 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: I don't know. Let's discuss it locally

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 17 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: This code computed width of text mask. Width text mask equal to sum of every glyph width.

**Comment by TolyaTalamanov** [comment]
@dmatveev Could you take another look at this ? Thank you!

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+        // Convert OpenCV scale to Freetype text height
+        auto sz = cv::getTextSize(text.text, FONT_HERSHEY_SIMPLEX, text.fs, 1, nullptr);
+        int font_height = static_cast<int>(sz.height * 3.0 / 2); // 3.0 / 2 for better aligning with OpenCV
+
+        FT_Set_Pixel_Sizes(m_face, font_height, font_height);
+    }
+
+    virtual ~FreeTypeBitmaskCreator() override
+    {
+        FT_Done_Face(m_face);
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: What if the face was invalid?

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+    FreeTypeBitmaskCreator(const std::string& path)
+    {
+        if (FT_Init_FreeType(&m_library) < 0)
+        {
+            util::throw_error(std::runtime_error("Failed to init FreeType library"));
+        }
+
+        if (FT_New_Face(m_library, path.c_str(), 0, &m_face))
+        {
+            util::throw_error(std::runtime_error("Failed to set font"));
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Throwing from constructor doesn't call destructor, so `m_library` leaks here.

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 25 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 26 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 27]
```diff
@@ -113,14 +104,22 @@ class FreeTypeBitmaskCreator : public IBitmaskCreator
 
     virtual ~FreeTypeBitmaskCreator() override
     {
-        FT_Done_Face(m_face);
-        FT_Done_FreeType(m_library);
+        GAPI_Assert(!FT_Done_Face(m_face));
+        GAPI_Assert(!FT_Done_FreeType(m_library));
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Never throw from destructor! That's harmful.

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 28]
```diff
@@ -113,14 +104,22 @@ class FreeTypeBitmaskCreator : public IBitmaskCreator
...
 
 private:
     FT_Library    m_library;
     FT_Face       m_face;
 
+    virtual void init() override
+    {
+        GAPI_Assert(!FT_Init_FreeType(&m_library));
+        GAPI_Assert(!FT_New_Face(m_library, m_font_path.c_str(), 0, &m_face));
+    }
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Extra init is unnecessary here

Comment (dmatveev) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: I mean, keep doing stuff in the constructor but more carefully

Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 27 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by dmatveev** [comment]
@TolyaTalamanov did you push an update?

**Comment by TolyaTalamanov** [comment]
@dmatveev  Yes

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 29]
```diff
@@ -31,10 +31,18 @@ class FreeTypeBitmaskCreator : public IBitmaskCreator
 public:
     FreeTypeBitmaskCreator(const std::string& path) : m_font_path(path)
     {
+        if (FT_Init_FreeType(&m_library) ||
+            FT_New_Face(m_library, m_font_path.c_str(), 0, &m_face))
+        {
+            m_is_initiliazed = false;
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: Noo, please.

Don't construct object (== `throw`) if the initialization has failed. Just it

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 30]
```diff
@@ -31,10 +31,18 @@ class FreeTypeBitmaskCreator : public IBitmaskCreator
...
+        }
     }
 
     virtual cv::Size computeMaskSize() override
     {
+        GAPI_Assert(m_is_initiliazed &&
+                    "FreeTypeBitmaskCreator is't initiliazed correctly !");
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: This is an anti-pattern

File: modules/gapi/src/api/freetype_mask_creator.hpp [hunk 31]
```diff
@@ -31,10 +31,18 @@ class FreeTypeBitmaskCreator : public IBitmaskCreator
 public:
     FreeTypeBitmaskCreator(const std::string& path) : m_font_path(path)
     {
+        if (FT_Init_FreeType(&m_library) ||
+            FT_New_Face(m_library, m_font_path.c_str(), 0, &m_face))
+        {
+            m_is_initiliazed = false;
+        }
```
Comment (dmatveev) [inline modules/gapi/src/api/freetype_mask_creator.hpp]: what you basically need is just
```c++
if (FT_Init_FreeType(&m_library) != 0)
{
   throw "Failed to initialize FT";
}
if (FT_New_Face(...) != 0)
{
   FT_Done_FreeType(...);
   throw "Failed to create a font face";
}
```

Please note I omitted an exception type for brevity but hope you'll address that.

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 29 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 30 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 31 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev

File: modules/gapi/src/api/freetype_mask_creator.hpp (hunk 31 above)
Comment (dmatveev) [reply modules/gapi/src/api/freetype_mask_creator.hpp]: ...waiting for an update :)

**Pull request closed without merging**